TAP-2.md

```diff
@@ -37,123 +37,84 @@ Provide all on-chain liquidity in a familiar limit orderbook interface to read (
-  priceImpact: string
+    fn encode_solution(
+        &self,
+        result: GetAmountOutResult,
+        sender: Address,
+        receiver: Address,
+    ) -> Result<RoutingSolution, Error>;
 }
-
-type DepthAtPrice = {
-  totalSize: string
-  averagePrice: string
-  route: Route
-  gasCost: Level['gasCost']
+```
+### 3. Price Level Calculator 
+Calculates discrete price levels for the orderbook using router simulations.
+```rust
+pub struct PriceLevel {
+    pub price: U256,
+    pub size: U256,
+    pub solution: RoutingSolution,
 }
 
-// Event System
-type OrderbookEvent = {
-  type: 'price_level_changed'
-  pair: TokenPair
-  timestamp: number
-  blockNumber: number
-  side: Side
-  price: string
-  newSize: string
-} | {
-  type: 'book_reset'
-  pair: TokenPair
-  timestamp: number
-  blockNumber: number
+pub struct PriceLevelCalculator {
+    router: Box<dyn Router>,
+    price_points: u32,
 }
-
-// Core Operations
-type Operations = {
-  getOrderbook: (pair: TokenPair, options?: OrderbookOptions) => Promise<OrderbookState>
-  
-  getQuoteForSize: (
-    pair: TokenPair,
-    size: string,
-    side: Side,
-    options?: OrderbookOptions
-  ) => Promise<SizeQuote>
-  
-  getSizeAtLimit: (
-    pair: TokenPair,
-    limitPrice: string,
-    side: Side,
-    options?: OrderbookOptions
-  ) => Promise<DepthAtPrice>
-  
-  findOptimalSize: (
-    pair: TokenPair,
-    targetPrice: string,
-    options?: FinetuneOptions
-  ) => Promise<SizeQuote>
-  
-  subscribeToEvents: (
+```
+### 4. Orderbook Manager
+Maintains orderbook state and handles updates as new protocol states arrive.
+```rust
+pub struct OrderbookState {
     pair: TokenPair,
-    onEvent: (event: OrderbookEvent) => void
-  ) => { unsubscribe: () => void }
-  
-  getSupportedTokens: () => Promise<string[]>
-  getSupportedPairs: () => Promise<TokenPair[]>
+    timestamp: u64,
+    block_number: u64,
+    bids: Vec<PriceLevel>,
+    asks: Vec<PriceLevel>,
+}
+pub struct OrderbookManager {
+    calculator: PriceLevelCalculator,
+    current_books: HashMap<TokenPair, OrderbookState>,
 }
 ```
+## Integration Flow
+1. Protocol Stream Manager connects to Tycho Indexer and receives real-time updates.
+2. Router maintains a cache of which pools are used in any solution for each orderbook.
+3. When pool states update:
+   - Router checks if any updated pools are used in any solution for a given orderbook.
+   - If yes, the entire orderbook needs to be recalculated as the optimal routes and prices at every amount level may have changed.
+4. For affected orderbooks:
+   - Every amount level must be recalculated.
+   - This is necessary because a change in any pool can affect which routes are optimal at all levels.
+   - PriceLevelCalculator generates a new set of price levels using the Router.
+## Extension Points
+1. Exchange simple router for more sophisticated (external) router.
+2. Local Fine-tuning
+3. Maintain Historical State (give users historical orderbook state)
```

Comment:
Very unclear what these mean exactly imo.

Reply:
Slightly pessimistic about using a more complex router here btw, because the amount of updates each block will start to increase exponentially per allowed swap depth, also things become interdependent between the individual orderbooks.